### PR TITLE
Add -with-file-output flag to coqdep, #5988

### PR DESCRIPTION
**Pull request by psteckler**
This flag takes a comma-delimited list of .v files, and for each file, creates the corresponding .v.d file.

Because coqdep does a traversal of `theories` and `plugins` on each invocation, passing many files in a single invocation should be much faster than calling it many times on a single file.

The main documentation for coqdep appears to be in the man page, I added an example there, though there are other things out of date there.

### CODE REVIEW

**Review by JasonGross — CHANGES_REQUESTED** [review]
Please update coq_makefile (maybe also Coq's built-in makefile) to take advantage of this flag

File: man/coqdep.1
Comment (psteckler) [reply man/coqdep.1]: What would the boolean flag indicate?

Comment (psteckler) [reply man/coqdep.1]: Oh, you mean take the files one at a time, with the flag indicating to output to a file

**Comment by psteckler** [comment]
@JasonGross the `-with-file-output` flag now takes no arguments, any following .v files on the command line will create .v.d files

I changed `coq_makefile`, adding a target `vdfiles` which will build .v.d files for all .v files the Makefile knows about. That replaces a multiple-target rule for particular .v.d files, such that coqdep was called for each .v.d file. Not sure if this is the right approach.

I think a change to the Coq Makefile to use this should be a separate PR, if this one is merged.

**Comment by psteckler** [comment]
To review this properly would probably involve running the code.

File: tools/coqdep_common.ml
Comment (JasonGross) [inline tools/coqdep_common.ml]: I think it might be better to make `coq_dependencies` simply take a boolean argument, determining for all files whether to output to stdout or to write .v.d files, and pass `!create_file_output` here, rather than in `treat_file`?

**Comment by psteckler** [comment]
I added the boolean flag to `coq_dependencies`. 

I also added the `-with-file-output` flag to `coqdep_boot`. On my laptop, the many invocations of `coqdep_boot` take about 5 seconds when building Coq. If this PR is accepted, we could change the Coq Makefile to turn that into a single invocation.

**Comment by psteckler** [comment]
With 8.7, I ran `make` on fiat-crypto, the `COQDEP` steps took about 6 minutes on my laptop.

With master, applying the changes in this PR, supplying the `-with-file-output` flag and all `.v` files on the command line, producing the `.v.d` files takes 1.5 seconds.

**Comment by maximedenes** [comment]
> With master, applying the changes in this PR, supplying the -with-file-output flag and all .v files on the command line, producing the .v.d files takes 1.5 seconds.

Impressive! And sounds like a very good idea. Did you check if it produced the same result? :)

**Comment by psteckler** [comment]
> Did you check if it produced the same result? :)

Yes, it does.

**Comment by psteckler** [comment]
@JasonGross Are there other changes to coq_makefile that are needed here (merging is blocked because of your review)?

**Review by JasonGross — CHANGES_REQUESTED** [review]
Argh, sorry, I wrote this a few days ago and forgot to click "submit review"

**Comment by gares** [comment]
I like the patch! One more thing worth doing is to remove the file if the file fails to parse or similar...
I.e. I'd like to get rid of this code https://github.com/coq/coq/blob/master/tools/CoqMakefile.in#L690
entirely, not just of the LHS.

File: tools/CoqMakefile.in
Comment (maximedenes) [reply tools/CoqMakefile.in]: > otherwise, you'll get in infinite loops, I think.

Can you explain how that could happen?

Comment (maximedenes) [reply tools/CoqMakefile.in]: You could try:
```
$(addsuffix .d,$(VFILES)): vdfiles
	$(SHOW)'COQDEP $(VFILES)'
	$(HIDE)$(COQDEP) $(COQLIBS) -dyndep var -c -with-file-output $(VFILES) $(if_ok)
```

Comment (JasonGross) [reply tools/CoqMakefile.in]: > Can you explain how that could happen?

`vdfiles` is declared `PHONY`, which means that it is considered outdated at the start of every run of `make`.  Because each `.v.d` is `include`d in the `Makefile`, `make` will check to see if the `.v.d` files are out of date.  Because they depend on an outdated target (`vdfiles`), they are, so they get remade.  When an included makefile is remade, I believe `make` restarts its processing of the entire file, in particular reconsidering all `PHONY` targets as outdated.  Then we are back to the starting condition, and we have an infinite loop.

> You could try:

No.  Then `make -jN` will invoke that rule `N` times in parallel, which is bad.

I think this will avoid the infinite loops, though, without needing to modify coqdep:
```makefile
$(addsuffix .d,$(VFILES)): $(VFILES) | vdfiles
```
As I understand it, order-only prerequisites are explicitly for this sort of case (where it's still a prerequisite, but its "outdated" status doesn't transfer to the parent rule).

Comment (psteckler) [reply tools/CoqMakefile.in]: These rules:
```
vdfiles: $(VFILES)
	$(SHOW)'COQDEP $^'
        $(HIDE)$(COQDEP) $(COQLIBS) -dyndep var -c -with-file-output $^

$(addsuffix .d,$(VFILES)): vdfiles
```
will call `coqdep` just once if you ask to build any number of `.v.d` files. But you get a "nothing to be done" message for each of them, which is annoying.

If you just run `make`, though, `make` gets called recursively, and `coqdep` gets called each time.

Comment (SkySkimmer) [reply tools/CoqMakefile.in]: With coq_makefile we have rules together for vo and glob, but if I do `make -j 2 foo.vo foo.glob` it doesn't run parallel coqc, why is that?
``` make
$(VOFILES): %.vo: %.v
	$(SHOW)COQC $<
	$(HIDE)$(TIMER) $(COQC) $(COQDEBUG) $(TIMING_ARG) $(COQFLAGS) $< $(TIMING_EXTRA)

# FIXME ?merge with .vo / .vio ?
$(GLOBFILES): %.glob: %.v
	$(TIMER) $(COQC) $(COQDEBUG) $(COQFLAGS) $<
```

Comment (psteckler) [reply tools/CoqMakefile.in]: Same if you add the order-only prereqs.

**Comment by psteckler** [comment]
I made the change @gares and @maximedenes mentioned today: write the output to a temp file, then rename it. If the temp file is empty, just delete it.

**Comment by psteckler** [comment]
I think the goal for the `coq_makefile`-generated Makefile should be, if you run `make`, it will run `coqdep` to generate all `.v.d` files. If you pass any number of `.v.d` targets, it should also run `coqdep` exactly once.

**Comment by psteckler** [comment]
Here's an oh-man-it's-complicated way to achieve the goal I mentioned:
```
# .v.d files made from .v files ###############################################                                                                                                                                                               
VDGOALS=$(strip $(filter %.v.d,$(MAKECMDGOALS)))
# build .v.d files if passed as targets                                                                                                                                                                                                       
ifneq ($(VDGOALS),)
$(VDGOALS) : $(VFILES)
        $(SHOW)'COQDEP $^'
        $(HIDE)$(COQDEP) $(COQLIBS) -dyndep var -c -with-file-output $(basename $@)
else
# build all .v.d files on recursive real-all invocation                                                                                                                                                                                       
ifeq ($(strip $(filter real-all,$(MAKECMDGOALS))),real-all)
vdfiles: $(VFILES)
        $(SHOW)'COQDEP $^'
        $(HIDE)$(COQDEP) $(COQLIBS) -dyndep var -c -with-file-output $^
$(addsuffix .d,$(VFILES)): vdfiles
endif
endif
```
If you pass `.v.d` targets, it builds just those ones, all at once, by invoking `coqdep` with the `with-file-output` flag. BTW, with the 8.7 `coq_makefile`, if you pass any `.v.d` target to the generated Makefile, it builds all possible `.v.d` files, one at a time. 

With this change, if you just run `make`, it will always build all the `.v.d` files, all at once. While that takes little time, it's kind of annoying.

**Comment by JasonGross** [comment]
> What's needed is a way to say, if any 

Here's a way, with a timestamp file, that works with `make -j2`, and doesn't rebuild the .v.d files on every run.
```makefile
VFILES := foo.v bar.v

.PHONY: all
all: $(VFILES:.v=.vo)
        echo 1

$(VFILES:.v=.vo) : %.vo : %.v
        cp $< $@
        sleep 1

.vdfiles: $(VFILES)
        touch $(addsuffix .d,$(VFILES)) && touch .vdfiles

$(addsuffix .d,$(VFILES)): .vdfiles

clean::
        rm -f $(VFILES:.v=.vo)
        rm -f $(addsuffix .d,$(VFILES)) .vdfiles

-include $(addsuffix .d,$(VFILES))
```

where `touch $(addsuffix .d,$(VFILES))` should be replaced with the call to `coqdep`

File: tools/CoqMakefile.in
Comment (JasonGross) [reply tools/CoqMakefile.in]: > Even without -jN, with that rule, if you try to build a particular .v.d file, coqdep gets run for every .v file.

I don't see that behavior.  You see two invocations of `coqdep` with this makefile?
```makefile
VFILES := foo.v bar.v

.PHONY: all
all: $(VFILES:.v=.vo)
        echo 1

$(VFILES:.v=.vo) : %.vo : %.v
        cp $< $@
        sleep 1

.PHONY: vdfiles
vdfiles:
        touch $(addsuffix .d,$(VFILES))

$(addsuffix .d,$(VFILES)): $(VFILES) | vdfiles

clean::
        rm -f $(VFILES:.v=.vo)
        rm -f $(addsuffix .d,$(VFILES))

-include $(addsuffix .d,$(VFILES))
```

Comment (JasonGross) [reply tools/CoqMakefile.in]: > With coq_makefile we have rules together for vo and glob, but if I do make -j 2 foo.vo foo.glob it doesn't run parallel coqc, why is that?

@SkySkimmer It does:
```
$ cat foo.v
Require Import Coq.ZArith.ZArith.

Time Definition foo0 := Eval vm_compute in Z.div_eucl.
$ cat _CoqProject
foo.v
$ coq_makefile -f _CoqProject -o Makefile3 && make -f Makefile3 clean && make -f Makefile3 -j2 foo.vo foo.glob
CLEAN
COQDEP foo.v
"coqc"  -q  -I .  -R . Top  foo.v
COQC foo.v
Finished transaction in 0.11 secs (0.095u,0.012s) (successful)
Finished transaction in 0.11 secs (0.092u,0.02s) (successful)
```

**Comment by psteckler** [comment]
> You see two invocations of coqdep with this makefile

No, sorry, I was looking at Maxime's suggestion. What I should have said is, you'll build all .v.d files for each one you request as a target.

**Comment by psteckler** [comment]
I added the timestamp-based target to build the `.v.d` files. The "real-all" target now has the `.v.d` files as a prerequisite, and cleaning removes those and the timestamp file.

File: tools/CoqMakefile.in
Comment (JasonGross) [inline tools/CoqMakefile.in]: I think you don't need this dependency; it's picked up automatically because the makefile `-include`s the `*.v.d` files.

File: tools/CoqMakefile.in
Comment (JasonGross) [inline tools/CoqMakefile.in]: I think you don't need to list `$(VDFILES)` here; it's covered by `$(ALLDFILES)` above.

**Review by JasonGross — APPROVED** [review]
I've suggested two minor changes, but I don't think either is essential (both are merely redundancies), and this looks good to me.  Thanks for doing this!

File: tools/CoqMakefile.in
Comment (psteckler) [reply tools/CoqMakefile.in]: Yup.

File: tools/CoqMakefile.in
Comment (psteckler) [reply tools/CoqMakefile.in]: I was not aware of this trick with `-include`, thanks.

**Comment by psteckler** [comment]
I changed the name of the tmp file from `_tmpNNNN` to `_coqdeptmpNNNN`, so if for whatever reason they're left around, it's clear where they came from.

**Comment by psteckler** [comment]
Don't merge this yet! `coqdep_boot` does not work correctly when given the file-output flag.

**Comment by maximedenes** [comment]
I'd like to have a final look at this patch before merging. Or @gares can have a look, of course.

**Comment by SkySkimmer** [comment]
What is the point of outputting to many small files as opposed to just doing `coqdep $allvfiles > vdeps.d` ?

**Comment by maximedenes** [comment]
Ok, I don't understand the design. I have exactly the same question as @SkySkimmer.
1. Why do you need a new flag?
2. Why many small files and a timestamp hack, rather than simply generating one file?

**Comment by psteckler** [comment]
1. coqdep writes its output to stdout, and the existing Makefile uses redirection to create files. The flag allows coqdep to write to files directly,

2. Maybe it's possible for the Makefile to use a single dependency file; I haven't looked into that. The goal with this PR was to create the same files as before.

**Comment by maximedenes** [comment]
> The flag allows coqdep to write to files directly, many at a time.

Ok, so writing a single file would solve this as well.

> Maybe it's possible for the Makefile to use a single dependency file; I haven't looked into that.

I thought the `.vdfiles` hack was there precisely because you needed in fact a single file. But maybe I misunderstood.

**Comment by psteckler** [comment]
> I thought the .vdfiles hack was there precisely because you needed in fact a single file.

The empty `.vdfiles` file means "all `.v.d` files have been written" and its timestamp can be compared with `.v` files to know if any `.v.d` files are out-of-date.

**Comment by SkySkimmer** [comment]
1 .d file approach https://github.com/SkySkimmer/coq/tree/coqdep-at-once https://github.com/SkySkimmer/coq/commit/f6d83bda8aff1b13465bbe242a790554654b3ad6
(the branch is called coqdep-at-once but for the main repository (as opposed to coq_makefile) it also does ocamldep in few commands (it needs 3*2=6 separate commands for plugins, checker and the rest, and for ml/mli and mllib/mlpack)).

**Comment by psteckler** [comment]
Writing a single file should be even faster.

Was there a reason for creating the individual files?

**Comment by JasonGross** [comment]
Apparently Coq Makefile used to create a single file.  This was changed in https://github.com/coq/coq/commit/615acfb1ab34c4352accfef438f21627d026e945 titled coq_makefile: dépendances + génération des fichiers html, with no apparent explanation.

I think we should go back to passing all the .v files to coqdep to generate a single .depends file, and just include that.

**Comment by SkySkimmer** [comment]
>1 .d file approach https://github.com/SkySkimmer/coq/tree/coqdep-at-once SkySkimmer/coq@f6d83bd

This breaks unimath because they want to process the .d files for stuff. Otherwise gitlab is passing it.

**Comment by psteckler** [comment]
> This breaks unimath because they want to process the .d files for stuff.

Could they change their build?

**Comment by psteckler** [comment]
So, what should happen with this PR?

**Comment by SkySkimmer** [comment]
I'm benching my commit at jenkins 241 to see how it goes.

**Comment by SkySkimmer** [comment]
SkySkimmer/coq@f6d83bd vs 2a857da2a88855a6c9f0fa7e48a8700c1613e0c7 (master when run)
https://ci.inria.fr/coq/view/benchmarking/job/benchmark-part-of-the-branch/241/console
```
┌──────────────────────────┬──────────────────────────┬───────────────────────────────────────┬────────────────────────────────────────┬─────────────────────────┬─────────────────┐
│                          │      user time [s]       │              CPU cycles               │            CPU instructions            │  max resident mem [KB]  │   mem faults    │
│                          │                          │                                       │                                        │                         │                 │
│             package_name │     NEW     OLD  PDIFF   │           NEW            OLD  PDIFF   │            NEW            OLD  PDIFF   │     NEW     OLD PDIFF   │ NEW OLD PDIFF   │
├──────────────────────────┼──────────────────────────┼───────────────────────────────────────┼────────────────────────────────────────┼─────────────────────────┼─────────────────┤
│            coq-fiat-core │  100.60  119.70 -15.96 % │  279032379770   315751328583 -11.63 % │   363445071048   424116612776 -14.31 % │  502232  502572 -0.07 % │  15   0  +nan % │
├──────────────────────────┼──────────────────────────┼───────────────────────────────────────┼────────────────────────────────────────┼─────────────────────────┼─────────────────┤
│          coq-fiat-crypto │ 3356.62 3716.68  -9.69 % │ 9355596792764 10341719205058  -9.54 % │ 15836566246788 17539300025959  -9.71 % │ 3282944 3282676 +0.01 % │   0   0  +nan % │
├──────────────────────────┼──────────────────────────┼───────────────────────────────────────┼────────────────────────────────────────┼─────────────────────────┼─────────────────┤
│         coq-fiat-parsers │  651.73  675.51  -3.52 % │ 1821538306849  1867845688099  -2.48 % │  2966511358622  3029504433708  -2.08 % │ 3535648 3534884 +0.02 % │   0   0  +nan % │
├──────────────────────────┼──────────────────────────┼───────────────────────────────────────┼────────────────────────────────────────┼─────────────────────────┼─────────────────┤
│   coq-mathcomp-odd_order │ 1362.21 1382.99  -1.50 % │ 3792155295519  3855828453697  -1.65 % │  6704276847509  6703997409949  +0.00 % │ 1121452 1109976 +1.03 % │   0   0  +nan % │
├──────────────────────────┼──────────────────────────┼───────────────────────────────────────┼────────────────────────────────────────┼─────────────────────────┼─────────────────┤
│               coq-geocoq │ 3056.72 3098.26  -1.34 % │ 8530157579011  8645086519808  -1.33 % │ 14290179855016 14317259972179  -0.19 % │ 1216268 1250012 -2.70 % │   0   0  +nan % │
├──────────────────────────┼──────────────────────────┼───────────────────────────────────────┼────────────────────────────────────────┼─────────────────────────┼─────────────────┤
│               coq-sf-vfa │   18.56   18.74  -0.96 % │   51833037991    51992857232  -0.31 % │    70938041264    71003227643  -0.09 % │  537372  537420 -0.01 % │   0   0  +nan % │
├──────────────────────────┼──────────────────────────┼───────────────────────────────────────┼────────────────────────────────────────┼─────────────────────────┼─────────────────┤
│    coq-mathcomp-fingroup │   59.07   59.64  -0.96 % │  163187831745   164356618411  -0.71 % │   229166960267   229639694955  -0.21 % │  530916  531244 -0.06 % │   0   0  +nan % │
├──────────────────────────┼──────────────────────────┼───────────────────────────────────────┼────────────────────────────────────────┼─────────────────────────┼─────────────────┤
│   coq-mathcomp-character │  267.10  269.26  -0.80 % │  742950126451   748274284417  -0.71 % │  1100695927681  1100851669853  -0.01 % │  981588  981696 -0.01 % │   0   0  +nan % │
├──────────────────────────┼──────────────────────────┼───────────────────────────────────────┼────────────────────────────────────────┼─────────────────────────┼─────────────────┤
│   coq-mathcomp-ssreflect │   38.05   38.34  -0.76 % │  105250679200   105625546166  -0.35 % │   130829168640   131452165963  -0.47 % │  477592  477404 +0.04 % │   0   0  +nan % │
├──────────────────────────┼──────────────────────────┼───────────────────────────────────────┼────────────────────────────────────────┼─────────────────────────┼─────────────────┤
│                coq-color │  566.43  570.58  -0.73 % │ 1586302193738  1595581304010  -0.58 % │  2004234150999  2015849829353  -0.58 % │ 1488188 1487940 +0.02 % │   0   0  +nan % │
├──────────────────────────┼──────────────────────────┼───────────────────────────────────────┼────────────────────────────────────────┼─────────────────────────┼─────────────────┤
│       coq-mathcomp-field │  448.65  451.86  -0.71 % │ 1247819523502  1258024855141  -0.81 % │  2095874074865  2096406956637  -0.03 % │  729044  728224 +0.11 % │   0   0  +nan % │
├──────────────────────────┼──────────────────────────┼───────────────────────────────────────┼────────────────────────────────────────┼─────────────────────────┼─────────────────┤
│     coq-mathcomp-algebra │  168.54  169.29  -0.44 % │  468733459198   471154710670  -0.51 % │   668518826832   668781608887  -0.04 % │  590200  590224 -0.00 % │   0   0  +nan % │
├──────────────────────────┼──────────────────────────┼───────────────────────────────────────┼────────────────────────────────────────┼─────────────────────────┼─────────────────┤
│ coq-mathcomp-real_closed │  166.24  166.94  -0.42 % │  462229342259   464153163350  -0.41 % │   711970726247   712398950777  -0.06 % │  731516  732284 -0.10 % │   0   0  +nan % │
├──────────────────────────┼──────────────────────────┼───────────────────────────────────────┼────────────────────────────────────────┼─────────────────────────┼─────────────────┤
│                  coq-vst │ 2177.01 2185.99  -0.41 % │ 6071022822166  6093105493201  -0.36 % │  8896882744656  8891900364981  +0.06 % │ 1951524 1951828 -0.02 % │   0   0  +nan % │
├──────────────────────────┼──────────────────────────┼───────────────────────────────────────┼────────────────────────────────────────┼─────────────────────────┼─────────────────┤
│    coq-mathcomp-solvable │  189.20  189.92  -0.38 % │  525885860360   528263563838  -0.45 % │   768026202487   768220029033  -0.03 % │  715068  714936 +0.02 % │   0   0  +nan % │
├──────────────────────────┼──────────────────────────┼───────────────────────────────────────┼────────────────────────────────────────┼─────────────────────────┼─────────────────┤
│           coq-coquelicot │   74.66   74.73  -0.09 % │  205598006569   205687035528  -0.04 % │   258842413511   258920649977  -0.03 % │  659968  660056 -0.01 % │   0   0  +nan % │
├──────────────────────────┼──────────────────────────┼───────────────────────────────────────┼────────────────────────────────────────┼─────────────────────────┼─────────────────┤
│               coq-sf-plf │   41.84   41.87  -0.07 % │  116552940203   116460303871  +0.08 % │   153002937171   153055790679  -0.03 % │  510672  510680 -0.00 % │   0   0  +nan % │
├──────────────────────────┼──────────────────────────┼───────────────────────────────────────┼────────────────────────────────────────┼─────────────────────────┼─────────────────┤
│              coq-bignums │   75.00   75.04  -0.05 % │  207726428237   208795928383  -0.51 % │   284421738640   285375454180  -0.33 % │  535420  535280 +0.03 % │  24   0  +nan % │
├──────────────────────────┼──────────────────────────┼───────────────────────────────────────┼────────────────────────────────────────┼─────────────────────────┼─────────────────┤
│                coq-flocq │   58.04   57.95  +0.16 % │  159054344225   158975740972  +0.05 % │   205329186602   205373779533  -0.02 % │  643768  643876 -0.02 % │ 161   0  +nan % │
├──────────────────────────┼──────────────────────────┼───────────────────────────────────────┼────────────────────────────────────────┼─────────────────────────┼─────────────────┤
│             coq-compcert │  802.95  801.36  +0.20 % │ 2239706838770  2234838124321  +0.22 % │  3413311591191  3415512000564  -0.06 % │ 1387948 1387876 +0.01 % │   0   0  +nan % │
├──────────────────────────┼──────────────────────────┼───────────────────────────────────────┼────────────────────────────────────────┼─────────────────────────┼─────────────────┤
│      coq-formal-topology │   33.84   33.66  +0.53 % │   93397098561    94179630472  -0.83 % │   125100821458   126873449235  -1.40 % │  482640  482564 +0.02 % │   0   0  +nan % │
├──────────────────────────┼──────────────────────────┼───────────────────────────────────────┼────────────────────────────────────────┼─────────────────────────┼─────────────────┤
│                coq-sf-lf │   15.54   15.43  +0.71 % │   42429693854    42418332415  +0.03 % │    59550787458    59568498344  -0.03 % │  419096  419124 -0.01 % │   0   0  +nan % │
└──────────────────────────┴──────────────────────────┴───────────────────────────────────────┴────────────────────────────────────────┴─────────────────────────┴─────────────────┘

```
-9.5% on fiat-crypto bringing it below the hour mark.

I'll make this into a proper PR later.

**Comment by mattam82** [comment]
I believe the idea of separate .d files is to incrementally update the dependencies during development but I wouldn't be too worried about that, adding files is a relatively rare operation in reality, and if a price has to be payed it should be at that time.

**Comment by Zimmi48** [comment]
Wasn't this PR superseded by #6217?

**Comment by SkySkimmer** [comment]
Yes.

**Pull request closed without merging**
